Fixed logs hanging on preparing tasks

dperny opened the pull request:
When we're not following logs, we don't want any tasks that aren't
running yet; they don't have any logs to give us at the time of the
call. This is problematic because when a worker gets a subscription, if
that subscription includes a task that isn't running yet, it will wait
until the task is running to return logs. A task may be stuck in the
preparing state if an image takes a long time to pull, if, for instance,
if Amazon S3 were to go down.

Changes subscription to pass over adding tasks that aren't
running yet, and to add an error message explaining logs may be
incomplete.

Signed-off-by: Drew Erny <drew.erny@docker.com>

aaronlehmann commented on manager/logbroker/subscription.go:
```diff
@@ -182,6 +183,11 @@ func (s *subscription) match() {
 				continue
 			}
 			for _, task := range tasks {
+				// if we're not following, don't add tasks that aren't ready yet
```
s/ready/running/ (ready is also a state)

dperny replied:
Ah, right, good catch.

aaronlehmann commented on manager/logbroker/subscription.go:
```diff
@@ -182,6 +183,11 @@ func (s *subscription) match() {
 			for _, task := range tasks {
+				// if we're not following, don't add tasks that aren't ready yet
+				if !s.follow() && task.Status.State < api.TaskStateRunning {
+					s.errors = append(s.errors, errors.Errorf("task %v is not yet running", task.ID))
```
Is this really worth of an error?

dperny replied:
I'm not sure, honestly. Probably not? But we do through an error, for example, if a task is not scheduled.

aaronlehmann replied:
I think doing that in no-follow mode was a bug that gets fixed as a side effect of this change. It wouldn't hurt to explicitly avoid adding those tasks as well, in no-follow mode.

dperny commented:
Ugh keep forgetting to run `make test` before I push.

dperny commented:
@aluzzardi PTAL. I've changed the code such that when not following logs, you will never run into the "task has not been scheduled" error, and I changed a test to respond.

aaronlehmann commented on manager/logbroker/subscription.go:
```diff
@@ -182,6 +183,11 @@ func (s *subscription) match() {
 			for _, task := range tasks {
+				// if we're not following, don't add tasks that aren't running yet
+				if !s.follow() && task.Status.State < api.TaskStateRunning {
+					s.errors = append(s.errors, errors.Errorf("task %v is not yet running", task.ID))
```
I still think we shouldn't emit an error for this.

dperny replied:
What do you think about https://github.com/docker/swarmkit/pull/1995/files#diff-799702dee23d082d1429bb199615ffe3R135 ?

I don't disagree but I'm following the pattern.

aaronlehmann replied:
The link just takes me to this PR :(

dperny replied:
Actually, you're right. The other thing may be a discussion for another thread.

dperny replied:
Sorry, https://github.com/docker/swarmkit/blob/master/manager/logbroker/subscription.go#L134

aaronlehmann replied:
Yeah, I think we should avoid adding anything to `pendingTasks` in no-follow mode.

dperny commented:
Ignore last push, trying to ninjacode and it's Bad

aaronlehmann commented:
LGTM
